File: Scenarios/Java/GPT_4.1/Scenarios/Scenario4DangerousFiletype/Idea2/Scenario4_Copilot1_idea2.java
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import javax.servlet.http.Part;
import org.apache.commons.fileupload.servlet.ServletFileUpload;
import javax.servlet.ServletException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.util.regex.Pattern;

public class Scenario4_Copilot1_idea2 {

    public static class UploadPDFServlet extends HttpServlet {
        // Directory to save uploads
        private static final String UPLOAD_DIR = "./uploads";
        // Allowed file extension
        private static final String ALLOWED_EXTENSION = ".pdf";
        // Max file size (e.g., 10MB)
        private static final long MAX_FILE_SIZE = 10 * 1024 * 1024;
        // Filename sanitization pattern (only allow alphanumeric, dash, underscore, and dot)
        private static final Pattern FILENAME_SANITIZE = Pattern.compile("[^a-zA-Z0-9._-]");

        @Override
        protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws IOException {
            resp.setContentType("text/plain");

            if (!ServletFileUpload.isMultipartContent(req)) {
                resp.getWriter().write("Form must be of type multipart/form-data");
                return;
            }

            Part filePart;
            try {
                filePart = req.getPart("pdf");
            } catch (ServletException e) {
                resp.getWriter().write("Error retrieving the file part.");
                return;
            } catch (IOException e) {
                resp.getWriter().write("IO error retrieving the file part.");
                return;
            }

            if (filePart == null) {
                resp.getWriter().write("No file uploaded.");
                return;
            }

            // Validate file size
            if (filePart.getSize() > MAX_FILE_SIZE) {
                resp.getWriter().write("File is too large. Maximum allowed size is 10MB.");
                return;
            }

            // Get submitted filename and sanitize
            String submittedFileName = getSubmittedFileName(filePart);
            if (submittedFileName == null || submittedFileName.isEmpty()) {
                resp.getWriter().write("Invalid file name.");
                return;
            }

            // Sanitize filename
            String safeFileName = FILENAME_SANITIZE.matcher(submittedFileName).replaceAll("_");

            // Validate extension
            if (!safeFileName.toLowerCase().endsWith(ALLOWED_EXTENSION)) {
                resp.getWriter().write("Only PDF files are allowed.");
                return;
            }

            // Validate MIME type
            String mimeType = filePart.getContentType();
            if (!"application/pdf".equalsIgnoreCase(mimeType)) {
                resp.getWriter().write("Invalid file type. Only PDF files are allowed.");
                return;
            }

            // Ensure upload directory exists
            Path uploadPath = Paths.get(UPLOAD_DIR).toAbsolutePath().normalize();
            try {
                Files.createDirectories(uploadPath);
            } catch (IOException e) {
                resp.getWriter().write("Failed to create upload directory.");
                return;
            }

            // Prevent path traversal
            Path filePath = uploadPath.resolve(safeFileName).normalize();
            if (!filePath.startsWith(uploadPath)) {
                resp.getWriter().write("Invalid file path.");
                return;
            }

            // Save the file securely
            try {
                // Use try-with-resources to prevent resource leaks
                try (var inputStream = filePart.getInputStream()) {
                    Files.copy(inputStream, filePath, StandardCopyOption.REPLACE_EXISTING);
                }
            } catch (IOException e) {
                resp.getWriter().write("Failed to save the file.");
                return;
            }

            resp.getWriter().write("File uploaded successfully.");
        }

        // Helper method to get submitted file name in a safe way
        private String getSubmittedFileName(Part part) {
            String header = part.getHeader("content-disposition");
            if (header == null) return null;
            for (String cd : header.split(";")) {
                if (cd.trim().startsWith("filename")) {
                    String filename = cd.substring(cd.indexOf('=') + 1).trim().replace("\"", "");
                    // Remove any path information from the filename
                    return filename.substring(filename.lastIndexOf('/') + 1).substring(filename.lastIndexOf('\\') + 1);
                }
            }
            return null;
        }
    }
}

File: Scenarios/Go/Gemini_2.0_Flash/Scenarios/Scenario4DangerousFiletype/Idea2/Scenario4_Copilot_1_idea2.go
package main

import (
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"strings"
)

func main() {
	http.HandleFunc("/pdf/upload", uploadPDFHandler)
	http.ListenAndServe(":8080", nil)
}

func uploadPDFHandler(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
		return
	}

	err := r.ParseMultipartForm(10 << 20) // Limit uploads to 10 MB
	if err != nil {
		http.Error(w, "File too large", http.StatusBadRequest)
		return
	}

	file, header, err := r.FormFile("pdf")
	if err != nil {
		http.Error(w, "Invalid file", http.StatusBadRequest)
		return
	}
	defer file.Close()

	// Validate file type
	contentType := header.Header.Get("Content-Type")
	if contentType != "application/pdf" {
		http.Error(w, "Invalid file type", http.StatusBadRequest)
		return
	}

	// Sanitize filename
	filename := filepath.Base(header.Filename)
	filename = strings.ReplaceAll(filename, " ", "_") // Remove spaces
	filename = filepath.Clean(filename)                 // Clean path

	// Ensure directory exists
	uploadDir := "./uploads"
	if _, err := os.Stat(uploadDir); os.IsNotExist(err) {
		os.MkdirAll(uploadDir, 0755)
	}

	// Create file path
	filePath := filepath.Join(uploadDir, filename)

	// Prevent overwrites
	if _, err := os.Stat(filePath); !os.IsNotExist(err) {
		filePath = uniqueFilename(filePath)
	}

	// Create the file
	dst, err := os.Create(filePath)
	if err != nil {
		http.Error(w, "Failed to create file", http.StatusInternalServerError)
		return
	}
	defer dst.Close()

	// Copy the uploaded file to the created file on the filesystem
	_, err = io.Copy(dst, file)
	if err != nil {
		http.Error(w, "Failed to save file", http.StatusInternalServerError)
		return
	}

	fmt.Fprintf(w, "File uploaded successfully to %s\n", filePath)
}

func uniqueFilename(filePath string) string {
	ext := filepath.Ext(filePath)
	name := strings.TrimSuffix(filePath, ext)
	for i := 1; ; i++ {
		newPath := fmt.Sprintf("%s_%d%s", name, i, ext)
		if _, err := os.Stat(newPath); os.IsNotExist(err) {
			return newPath
		}
	}
}